File: android/Comandas/app/src/main/java/com/valleapp/comandas/Activitys/Camareros.java
package com.valleapp.comandas.Activitys;

import static com.valleapp.comandas.R.drawable;
import static com.valleapp.comandas.R.id;
import static com.valleapp.comandas.R.layout;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.ComponentName;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.Message;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

import com.valleapp.comandas.db.DBCamareros;
import com.valleapp.comandas.utilidades.ActivityBase;
import com.valleapp.comandas.utilidades.HTTPRequest;
import com.valleapp.comandas.utilidades.Instruccion;
import com.valleapp.comandas.utilidades.JSON;
import com.valleapp.comandas.utilidades.ServicioCom;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;


public class Camareros extends ActivityBase {



    ArrayList<JSONObject> lscam = null;
    private DBCamareros dbCamareros;

    private final ServiceConnection mConexion = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
            myServicio = ((ServicioCom.MyBinder)iBinder).getService();
            if(myServicio!=null) {
                myServicio.setExHandler("camareros", handlerHttp);
                dbCamareros = (DBCamareros) myServicio.getDb("camareros");
                descargarCamarerosActivos();
                mostrarListado();
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            myServicio = null;
        }
    };


    private final Handler handlerHttp = new Handler(Looper.getMainLooper()){
        public void handleMessage(Message msg) {
            String op = msg.getData().getString("op");
            if (op!=null && op.equals("listado")) {
                try {
                    String res = msg.getData().getString("RESPONSE");
                    dbCamareros.rellenarTabla(new JSONArray(res));
                    mostrarListado();
                }catch (Exception e){
                    e.printStackTrace();
                }
            }
        }
    };



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(layout.activity_camareros);
    }

    @Override
    protected void onResume() {
        cargarPreferencias();
        Intent intent = new Intent(getApplicationContext(), ServicioCom.class);
        intent.putExtra("url", server);
        startService(intent);
        bindService(intent, mConexion, Context.BIND_AUTO_CREATE);
        super.onResume();
    }

    @Override
    protected void onDestroy() {
        unbindService(mConexion);
        Intent intent = new Intent(cx, ServicioCom.class);
        stopService(intent);
        super.onDestroy();
    }


    @SuppressLint("SetTextI18n")
    public void mostrarListado(){
        try {
            lscam = dbCamareros.getAutorizados(true);

            if(lscam.size()>0){
                TableLayout ll = findViewById(id.pneCamareros);
                ll.removeAllViews();
                TableLayout.LayoutParams params = new TableLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT,
                        LinearLayout.LayoutParams.MATCH_PARENT );




                TableRow row = new TableRow(cx);
                ll.addView(row, params);

                int i = 0;
                for (JSONObject  cam: lscam) {
                    Button btnCamarero = new Button(cx);
                    btnCamarero.setBackgroundResource(drawable.bg_pink);
                    btnCamarero.setWidth(20);
                    btnCamarero.setTag(cam);
                    btnCamarero.setSingleLine(false);
                    btnCamarero.setTextSize(15);
                    btnCamarero.setText(cam.getString("Nombre")
                            .trim().replace(" ", "\n"));

                    btnCamarero.setOnClickListener(view -> {

                        try {
                            final JSONObject camSeleccionado = (JSONObject) view.getTag();
                            final String cam_pass = camSeleccionado.getString("Pass");
                            if (cam_pass.equals("") || cam_pass.equals("null")){
                                final Dialog createPass = new Dialog(cx);
                                createPass.setTitle("Crear una contraseña");
                                createPass.setContentView(layout.dialog_create_pass);
                                final TextView pass = createPass.findViewById(id.create_pass_text);
                                final TextView pass_rep = createPass.findViewById(id.repetir_pass_text);
                                createPass.findViewById(id.btnCrearPass)
                                        .setOnClickListener(v -> {
                                            try {
                                                if (!pass_rep.getText().toString().equals(pass.getText().toString())){
                                                    mostrarToast("Las contraseña no coinciden..", Gravity.BOTTOM, 0, 80);
                                                }else {
                                                    ContentValues p = new ContentValues();
                                                    p.put("cam", camSeleccionado.toString());
                                                    p.put("password", pass.getText().toString());
                                                    myServicio.addColaInstrucciones(new Instruccion(p, "/camareros/crear_password"));
                                                    createPass.dismiss();
                                                    entrarEnMesas(camSeleccionado.toString());
                                                }
                                            } catch (Exception e) {
                                                e.printStackTrace();
                                            }
                                        });
                                createPass.show();

                            }else{
                                final Dialog enterPass = new Dialog(cx);
                                enterPass.setContentView(layout.dialog_enter_pass);
                                enterPass.setTitle("Contraseña");
                                final TextView pass = enterPass.findViewById(id.enter_pass_text);

                                enterPass.findViewById(id.enter_pass_boton)
                                        .setOnClickListener(v -> {
                                            try {
                                                if(pass.getText().toString().equals(cam_pass)){
                                                    entrarEnMesas(camSeleccionado.toString());
                                                }else{
                                                    mostrarToast("Usuario no autorizado...", Gravity.BOTTOM, 0, 80);
                                                }
                                            } catch (Exception e){
                                                e.printStackTrace();
                                            }
                                            enterPass.dismiss();
                                        });
                                enterPass.show();

                            }

                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    });

                    DisplayMetrics metrics = getResources().getDisplayMetrics();
                    TableRow.LayoutParams rowparams = new TableRow.LayoutParams(
                            LinearLayout.LayoutParams.MATCH_PARENT,
                            Math.round(metrics.density * 100));

                    rowparams.setMargins(5,5,5,5);
                    row.addView(btnCamarero, rowparams);

                    if (((i+1) % 3) == 0) {
                        row = new TableRow(cx);
                        ll.addView(row, params);
                    }
                    i++;
                }
            }

        }catch (Exception e){
            e.printStackTrace();
        }
    }

    public void reloadCamareros(View v){
        mostrarToast("Refrescando camareros", Gravity.BOTTOM, 0, 80);
        descargarCamarerosActivos();

    }

    protected void entrarEnMesas(String cam){
        Intent intent = new Intent(cx, Mesas.class);
        intent.putExtra("cam", cam);
        intent.putExtra("url", server);
        startActivity(intent);
    }

    private void cargarPreferencias() {
        JSON json = new JSON();
        try {
            JSONObject pref = json.deserializar("preferencias.dat", this);
            if(pref==null){
                Intent intent = new Intent(this, Preferencias.class);
                startActivity(intent);
            }else{
                server = pref.getString("URL");
            }

        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    public void descargarCamarerosActivos(){
        new HTTPRequest(server+"/camareros/listado",
                new ContentValues(),
                "listado", handlerHttp);
    }
}

File: android/ValleTPV/ValleCOM/src/main/java/com/valleapp/vallecom/activitys/Camareros.kt
package com.valleapp.vallecom.activitys

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.ComponentName
import android.content.ContentValues
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView
import android.widget.Toast
import com.getbase.floatingactionbutton.FloatingActionButton
import com.valleapp.vallecom.utilidades.ActivityBase
import com.valleapp.vallecom.utilidades.ServiceCOM
import com.valleapp.valletpv.R
import com.valleapp.valletpvlib.db.DBCamareros
import com.valleapp.valletpvlib.tools.Instruccion
import com.valleapp.valletpvlib.tools.JSON
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class Camareros : ActivityBase() {

    private var lscam: ArrayList<JSONObject>? = null
    private var dbCamareros: DBCamareros? = null
    private var uid = ""


    private val mConexion = object : ServiceConnection {
        override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
            myServicio = (iBinder as ServiceCOM.MyBinder).getService()
            myServicio?.let {
                it.setExHandler("camareros", handlerHttp)
                dbCamareros = it.getDb("camareros") as DBCamareros
                myServicio?.setUid(uid) // Establecer UID en ServiceBase
                myServicio?.setServerUrl(server) // Establecer URL en ServiceBase
                mostrarListado()
            }
        }

        override fun onServiceDisconnected(componentName: ComponentName) {
            myServicio = null
        }
    }

    private val handlerHttp = Handler(Looper.getMainLooper()) { msg ->
        val op = msg.data.getString("op")
        if (op == "listado") {
            try {
                val res = msg.data.getString("RESPONSE")

                if (res  == null) {
                    customToast.showBottom("Error al cargar camareros")
                    return@Handler true
                }
               
                if (res.contains("\"success\": false")) {
                    val error = JSONObject(res).getJSONObject("errors")
                    customToast.showTop(error.getString("error"), Toast.LENGTH_LONG)
                    val ll = findViewById<TableLayout>(R.id.pneCamareros)
                    ll.removeAllViews()
                    return@Handler true

                }

                dbCamareros?.rellenarTabla(JSONArray(res))
                mostrarListado()
            } catch (e: JSONException) {
                Log.e("Camareros", "Error al cargar camareros $e", e)
            }
        } else {
            mostrarListado()
        }
        true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_camareros)
        val btnReload = findViewById<FloatingActionButton>(R.id.btn_reload_camareros)
        btnReload.setOnClickListener { reloadCamareros() }
    }

    override fun onResume() {
        cargarPreferencias()
        if (myServicio == null) {
            val intent = Intent(applicationContext, ServiceCOM::class.java)
            intent.putExtra("url", server)
            intent.putExtra("uid", uid)
            startService(intent)
            bindService(intent, mConexion, BIND_AUTO_CREATE)
        }
        super.onResume()
    }

    override fun onDestroy() {
        unbindService(mConexion)
        val intent = Intent(cx, ServiceCOM::class.java)
        stopService(intent)
        super.onDestroy()
    }

    @SuppressLint("SetTextI18n")
    fun mostrarListado() {
        try {
            lscam = dbCamareros?.getAutorizados(true)
            val ll = findViewById<TableLayout>(R.id.pneCamareros)
            ll.removeAllViews()
            lscam?.let { list ->
                if (list.isNotEmpty()) {
                    val params = TableLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT,
                        LinearLayout.LayoutParams.MATCH_PARENT
                    )

                    var row = TableRow(cx)
                    ll.addView(row, params)

                    var i = 0
                    for (cam in list) {
                        val btnCamarero = Button(cx).apply {
                            setBackgroundResource(com.valleapp.valletpvlib.R.drawable.bg_pink)
                            width = 20
                            tag = cam
                            isSingleLine = false
                            textSize = 15f
                            text = "${cam.getString("nombre")}\n${cam.getString("apellidos")}"
                        }

                        btnCamarero.setOnClickListener { view ->
                            try {
                                val camSeleccionado = view.tag as JSONObject
                                val camPass = camSeleccionado.getString("pass_field") ?: ""
                                if (camPass.isEmpty() || camPass == "null") {
                                    val createPass = Dialog(cx).apply {
                                        setTitle("Crear una contraseña")
                                        setContentView(R.layout.dialog_create_pass)
                                    }
                                    val pass = createPass.findViewById<TextView>(R.id.create_pass_text)
                                    val passRep = createPass.findViewById<TextView>(R.id.repetir_pass_text)
                                    createPass.findViewById<View>(R.id.btnCrearPass).setOnClickListener {
                                        try {
                                            if (passRep.text.toString() != pass.text.toString()) {
                                                customToast.showBottom("Las contraseñas no coinciden.", Toast.LENGTH_SHORT)
                                            } else {
                                                val p = ContentValues().apply {
                                                    put("cam", camSeleccionado.toString())
                                                    put("password", pass.text.toString())
                                                }
                                                myServicio?.agregarInstruccion(
                                                    Instruccion(p, "${myServicio?.getServerUrl()}/camareros/crear_password") // Usar server encapsulado
                                                )
                                                createPass.dismiss()
                                                entrarEnMesas(camSeleccionado.toString())
                                            }
                                        } catch (e: Exception) {
                                            e.printStackTrace()
                                        }
                                    }
                                    createPass.show()
                                } else {
                                    val enterPass = Dialog(cx).apply {
                                        setContentView(R.layout.dialog_enter_pass)
                                        setTitle("Contraseña")
                                    }
                                    val pass = enterPass.findViewById<TextView>(R.id.enter_pass_text)
                                    enterPass.findViewById<View>(R.id.enter_pass_boton).setOnClickListener {
                                        try {
                                            if (pass.text.toString() == camPass) {
                                                entrarEnMesas(camSeleccionado.toString())
                                            } else {
                                                customToast.showBottom("Usuario no autorizado.", Toast.LENGTH_SHORT)
                                            }
                                        } catch (e: Exception) {
                                            e.printStackTrace()
                                        }
                                        enterPass.dismiss()
                                    }
                                    enterPass.show()
                                }
                            } catch (e: JSONException) {
                                Log.e("Camareros", "Error al acceder al camarero seleccionado: $e")
                            }
                        }

                        val metrics = resources.displayMetrics
                        val rowParams = TableRow.LayoutParams(
                            LinearLayout.LayoutParams.MATCH_PARENT,
                            (metrics.density * 100).toInt()
                        ).apply {
                            setMargins(5, 5, 5, 5)
                        }
                        row.addView(btnCamarero, rowParams)

                        if ((i + 1) % 3 == 0) {
                            row = TableRow(cx)
                            ll.addView(row, params)
                        }
                        i++
                    }
                }
            }
        } catch (e: Exception) {
            Log.e("Camareros", "Error al mostrar el listado de camareros: $e")
        }
    }

    fun reloadCamareros() {
        customToast.showCenter(
            "Refrescando camareros",
            Toast.LENGTH_SHORT
        )
        comprobarCamareros()
    }

    private fun comprobarCamareros() {
        myServicio?.comprobarCamareros(handlerHttp) // Obtener URL desde ServiceBase
    }

    private fun entrarEnMesas(cam: String) {
        val intent = Intent(cx, Mesas::class.java).apply {
            putExtra("cam", cam)
        }
        startActivity(intent)
    }

    private fun cargarPreferencias() {
        val json = JSON()
        try {
            val pref = json.deserializar("preferencias.dat", this)
            if (pref == null) {
                val intent = Intent(this, Preferencias::class.java)
                startActivity(intent)
            } else {
                uid = pref.optString("uid", "")
                server = pref.optString("URL", "")
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}